email/exporter.go

```diff
@@ -0,0 +1,159 @@
+	maxConcurrentRequests int
+	limiter               *rate.Limiter
+	toSend                []string
+	client                PardotClient
+	emailsHandledCounter  prometheus.Counter
+	log                   blog.Logger
+}
+
+var _ emailpb.ExporterServer = (*ExporterImpl)(nil)
+
+// NewExporterImpl initializes an ExporterImpl with the given client and
+// configuration. Both perDayLimit and maxConcurrentRequests should be
+// distributed proportionally among instances based on their share of the daily
+// request cap. For example, if the total daily limit is 50,000 and one instance
+// is assigned 40% (20,000 requests), it should also receive 40% of the max
+// concurrent requests (e.g., 2 out of 5). For more details, see:
+// https://developer.salesforce.com/docs/marketing/pardot/guide/overview.html?q=rate%20limits
+func NewExporterImpl(client PardotClient, perDayLimit float64, maxConcurrentRequests int, scope prometheus.Registerer, logger blog.Logger) *ExporterImpl {
+	limiter := rate.NewLimiter(rate.Limit(perDayLimit/86400.0), maxConcurrentRequests)
+
+	emailsHandledCounter := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "email_exporter_emails_handled",
+		Help: "Total number of emails handled by the email exporter",
+	})
+	scope.MustRegister(emailsHandledCounter)
+
+	impl := &ExporterImpl{
+		maxConcurrentRequests: maxConcurrentRequests,
+		limiter:               limiter,
+		toSend:                make([]string, 0, contactsQueueCap),
+		client:                client,
+		emailsHandledCounter:  emailsHandledCounter,
+		log:                   logger,
+	}
+	impl.wake = sync.NewCond(&impl.Mutex)
+
+	queueGauge := prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "email_exporter_queue_length",
+		Help: "Current length of the email export queue",
+	}, func() float64 {
+		impl.Lock()
+		defer impl.Unlock()
+		return float64(len(impl.toSend))
+	})
+	scope.MustRegister(queueGauge)
+
+	return impl
+}
+
+// SendContacts enqueues the provided email addresses. If the queue cannot
+// accommodate the new emails, an ErrQueueFull is returned.
+func (impl *ExporterImpl) SendContacts(ctx context.Context, req *emailpb.SendContactsRequest) (*emptypb.Empty, error) {
+	if core.IsAnyNilOrZero(req, req.Emails) {
+		return nil, berrors.InternalServerError("Incomplete UpsertEmails request")
+	}
+
+	impl.Lock()
+	defer impl.Unlock()
+
+	spotsLeft := contactsQueueCap - len(impl.toSend)
+	if spotsLeft < len(req.Emails) {
+		return nil, ErrQueueFull
+	}
+	impl.toSend = append(impl.toSend, req.Emails...)
+	// Wake waiting workers to process the new emails.
+	impl.wake.Broadcast()
+
+	return &emptypb.Empty{}, nil
+}
+
+// Start begins asynchronous processing of the email queue. When the parent
+// daemonCtx is cancelled the queue will be drained and the workers will exit.
+func (impl *ExporterImpl) Start(daemonCtx context.Context) {
+	go func() {
+		<-daemonCtx.Done()
+		impl.Lock()
+		defer impl.Unlock()
+		// Wake waiting workers to exit.
+		impl.wake.Broadcast()
```

Comment:
FWIW, https://pkg.go.dev/sync#Cond.Broadcast:

> It is allowed but not required for the caller to hold c.L during the call.

Given that it's not required, I'd skip the locking here for simplicity.